### ISSUE TITLE
[UX] Disambiguate "Layouts" vs "Layout templates" in the admin interface.

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#3607]
I noticed that the tab on the Layouts page is labeled `Install new Layouts` when it should really be `Install new Layout TEMPLATES` as per our new disambiguation between `Layouts` and `Layout templates`. 
<img width="830" alt="Screen Shot 2019-03-21 at 5 43 32 PM" src="https://user-images.githubusercontent.com/397895/54789608-12775700-4c01-11e9-821a-52c86c623166.png">

---
~PR by @seancav03 https://github.com/backdrop/backdrop/pull/2599~
Updated PR https://github.com/backdrop/backdrop/pull/3278 (includes commit above)

**Comment by klonos** [comment]
To be absolutely exact...

- "Install new layouts" should be "Install new layout templates"
- "Update layouts" should be "Update layout templates" (but should really go entirely #2714)
- "Settings" should really be "List layout templates" or "Available layout templates" (because people are not just viewing a list in that page; they can also enable/disable templates)

In theory we should start calling the contrib projects "Layout templates" as well (and rename things in the b.org menu structure, and documentation/references too).

**Comment by jenlampton** [comment]
The settings page may have more settings than just allowed layout
templates, so settings is fine there. Others tho, :thumbs up:

**Comment by BWPanda** [comment]
In writing https://api.backdropcms.org/documentation/overriding-layout-templates, I struggled knowing what to call these things (for lack of a better word). 'Layouts' seems to be obvious, because that's the directory name where they're stored, and 'templates' might refer to the *.tpl.php files in those layouts, but then I saw this and was like "Uh...what!?"

![layouts](https://user-images.githubusercontent.com/2385329/65820450-e41ea800-e231-11e9-8ac9-d9b82366b89a.png)

I thought 'Moscone Flipped' was the layout, 'cause it's stored in `/core/layouts/moscone_flipped`; but 'Default layout' is the layout and 'Moscone Flipped' is the template...? Argh!

**Comment by jenlampton** [comment]
Yeah, it's confusing. A "Layout" is a specific path (or path pattern like `node/%`) that contains both block placements and configuration options, including which `Layout template` is in use. Example: `This is my Home Page layout`. (It could be 2 columns, stripes, custom, etc)

**Comment by stpaultim** [comment]
I think this PR looks good.

**Comment by klonos** [comment]
I'm not sure about `Install layout projects` 🤔 ...we are not calling the rest `module projects` and `theme projects`. I would still like to refer to those as "layout templates". The way I see it, we have 4 sorts of them:

- core ones: the ones that come with Backdrop OOTB, under `/core/layouts`
- contrib ones: the ones you download from b.org or GitHub, or install via the Project Browser
- custom ones: the ones people can create via code, and put under the root `/layouts` folder
- "flexible" ones: the custom ones people can create via the admin UI, `/admin/structure/layouts/settings/flexible-template/add`

**Comment by stpaultim** [comment]
Adding a screenshot for reference. 

![layouts-test-1284](https://user-images.githubusercontent.com/3144571/91758966-75ab3e00-eb96-11ea-9b4d-b823c3c912fa.jpg)

> (contrib and custom layout template "projects" may include multiple templates in them, the same way that module "projects" can have multiple sub-modules)

This partially highlights the problem I think. Flexible layout templates can not come bundled together, nor can the user find them in a convenient directory in the file system. They are nothing but config. 

How do we distinguish between a configured layout template and one that can be downloaded through the installer? Not 100% sure this distinction is important, but I'm leaning towards thinking that it is at least helpful. The downside is that "Layout Projects" is a bit more cumbersome (but not too bad).

**Comment by jenlampton** [comment]
> I would still like to refer to those as "layout templates". 

Even though a "layout project" contains multiple "layout templates"?

> How do we distinguish between a configured layout template and one that can be downloaded through the installer?

by "Flexible layout templates" vs "layout templates".

I thought these terms were decided on [long ago](https://github.com/backdrop/backdrop-issues/issues/3607#issuecomment-536212057)...  you both agreed on these terms back then...

**Comment by stpaultim** [comment]
@jenlampton - Just to be clear (and my comment was anything but clear), I am mostly in agreement with you and your proposal. I was acknowledging @klonos point about the different treatment with modules and themes, but trying to point out that there is a reason for it. 

A flexible layout template would never have a project, but a core or contrib layout template would. I agree that is probably a helpful distinction.

**Comment by klonos** [comment]
Sorry for the late reply. I would like to get this moving again.

> edit: sub-modules are also (almost always) part of a single module. The code is in a sub-folder within. ...

Although it's best practice to include submodules in subfolders, that is _not_ a requirement, and also not _always_ the case. The "main" or "primary" module needs to have the same name for its .module and .info file as the main folder containing all submodules of a project, but that's just about it when it comes to requirements. In fact, there are projects like https://github.com/backdrop-contrib/ubercart where the main module is nothing more than a "placeholder module", in order to satisfy that requirement - there's no actual code in the .module file at all.

> ...The features add to what the main module provides. It's not really the same thing with layout templates.

Also not always the case. On top of it, you seem to be thinking from a developer point of view here @jenlampton, and then exposing developer-focused concepts/terminology in the UI. Chances are that people using the admin UI exclusively will never need to know the (sub)folder structure, and won't care about it either - they'll just want to use the templates that are bundled in a single layout template project.

> Even though a "layout project" contains multiple "layout templates"?

Yes. Same way that a single "module" (project) can contain multiple (sub)modules, yet we don't call the singular bundle of modules a "module project" nor using the plural "modules" to refer to it, I don't think that we should be calling a "layout template" (project) a "layout project". In fact, I would argue that calling it that makes things even more confusing, as the name still implies that this project type holds/provides _layouts_ - it doesn't. It holds one or several _layout template(s)_, and that is what we are trying to disambiguate here.

So to sum it up, here's what I think would make sense from an end user point of view:
![image](https://user-images.githubusercontent.com/2423362/211667977-ff73e495-30b7-407c-868e-79bf8bcadd43.png)
![image](https://user-images.githubusercontent.com/2423362/211668143-ea65cfe0-7d3e-409c-823c-f87658d23ff4.png)
![image](https://user-images.githubusercontent.com/2423362/211668254-ee7374c0-13ad-4b27-afaa-8dcf4b01df1f.png)

Alternative PR up for testing/review/comparison: https://github.com/backdrop/backdrop/pull/4303

@jenlampton if you still insist on calling these "layout projects", can you please rebase your PR, so that people can play around with the UI and compare the two alternatives? Thanks.

**Comment by jenlampton** [comment]
I do think it will be better if people understand that they can't install their `Wheelan` Layout template without installing the `Radix` Layout project. Using the same name for both things is going to be slightly more confusing.

I don't feel that strongly about "Layout project" vs "Layout template", I was just surprised that had we agreed on the names and then when I implemented them, we wanted something else.

Any of these PRs is still better than what we have now!  :)

**Comment by yorkshire-pudding** [comment]
I had a look in the core folders for layout and the only one with more than one template in is legacy. I half expected moscone and moscone_flipped to be two templates in one folder but no, they each have a separate folder with a separate .info file. I'm quite late in the day to this conversation, but I've had a look through and layout template feels a more intuitive form of words for what they are.

**Comment by klonos** [comment]
> I was just surprised that had we agreed on the names...

I don't think have never agreed. See: https://github.com/backdrop/backdrop-issues/issues/3607#issuecomment-475928006 and https://github.com/backdrop/backdrop-issues/issues/3607#issuecomment-683823073 😅

**Comment by klonos** [comment]
@yorkshire-pudding thanks for taking the time to review and provide feedback 🙏🏼 

Yeah, renaming the `/layouts` and `/core/layouts` folders and reorganizing the structure within them is a whole other can of worms 😓  (and we're a bit late to make that change, even if we wanted to - it'll break lots of things, so maybe a 2.x thing ...may just leave it alone). Same goes for the `type = layout` lines in .info files. I think that developers can live with the fact that layout template projects are simply called "layouts" (or to be more precise they have `layout` for the machine name of their type) 🤷🏼‍♂️ 

The goal here is to make it less confusing in more obvious places and user-facing text, which includes this PR here, changing what we call these in our documentation and inline comments, and perhaps renaming certain things in b.org.

**Comment by jenlampton** [comment]
> I had a look in the core folders for layout and the only one with more than one template in is legacy. I half expected moscone and moscone_flipped to be two templates in one folder but no, they each have a separate folder with a separate .info file.

All layout templates need to have an `.info` file, so it would be impossible to have any two layouts share one. Looking at core isn't a great example. We don't have very many core modules with submodules either, I can think only of the field modules that do that (`number`, `text`, etc). 

In most cases the project you download from backdropcms.org is going to contain a set of layout templates that all use the same layout framework (something like bootstrap, zen, etc) but have different configurations of rows and columns. 

> Yeah, renaming the /layouts and /core/layouts folders and reorganizing the structure within them is a whole other can of worms 😓 (and we're a bit late to make that change, even if we wanted to

Oi, you mean renaming them to `layout-templates`?  I hadn't thought of that :/

> if you still insist on calling these "layout projects", can you please rebase your PR, so that people can play around with the UI and compare the two alternatives? Thanks.

I did end up rebasing my PR: https://github.com/backdrop/backdrop/pull/4304 but I expect we'll probably use yours. I did a quick code review and found a few unnecessary comment changes, but otherwise it looks good!

**Comment by klonos** [comment]
> Oi, you mean renaming them to `layout-templates`?  I hadn't thought of that :/

Yes, either `layout-templates` or `layout_templates` ...either way, it's not ideal DX to have a thing called "layouts" which doesn't make sense, since layouts are config .json files that is kept under the config folders - the `/layouts` and `/core/layouts` folders hold layout templates, but as you can understand and as I said, whole different can of worms there 😓 

> I did a quick code review (of your PR) and found a few unnecessary comment changes, but otherwise it looks good!

Thanks. Reverted those 👍🏼

**Comment by klonos** [comment]
@jenlampton @stpaultim @yorkshire-pudding you've all chimed in here recently, indicating that you like where this is heading, however have you actually had the chance to test or review the code? I would really like this in for 1.24.0 (and 1.23.2), so pwetty pwease 🙏🏼 🙂

**Comment by yorkshire-pudding** [comment]
@klonos - which one should I look to test/review?

**Comment by klonos** [comment]
@yorkshire-pudding mine is this: https://github.com/backdrop/backdrop/pull/4303 ...but in general, hovering over a PR link in the sidebar should reveal a pop-up which indicates the name of the repo it's coming from:
![image](https://user-images.githubusercontent.com/2423362/212655729-c8d0507a-90e1-472b-bf68-3c89593c813e.png)

**Comment by yorkshire-pudding** [comment]
Thanks @klonos - I've tested this in the sandbox and all looks good and I think makes more sense than what is there currently.  I've also reviewed the code and think it looks fine.

**Comment by BWPanda** [comment]
I reviewed @klonos' PR and tested in the sandbox. Looks good to me. I'm not sure about @jenlampton's 'Layout project' terminology... Perhaps something to discuss in a new issue. FTR, if necessary, I'd prefer something like 'Layout bundle'.

I've merged https://github.com/backdrop/backdrop/pull/4303 into 1.x and 1.24.x. THanks to @klonos, @jenlampton, @stpaultim & @yorkshire-pudding for their feedback here.
